lib/view_model/active_record/update_operation.rb

```diff
@@ -526,35 +540,121 @@ def build_updates_for_has_many_through(association_data, reference_strings, upda
+              {
+                ref:           ref,
+                viewmodel_ref: update_context.resolve_reference(ref).viewmodel_reference
+              }
+            end
+
+            if fupdate.before || fupdate.after
+              moved_refs =
+                fupdate.values.map { |ref| update_context.resolve_reference(ref).viewmodel_reference }.to_set
+
+              target_indirect_viewmodels =
+                target_indirect_viewmodels.reject { |target| moved_refs.include?(target[:viewmodel_ref]) }
+
+              ref   = fupdate.before || fupdate.after
+              index = target_indirect_viewmodels.find_index { |target| target[:viewmodel_ref] == ref }
+              unless index
+                raise ViewModel::DeserializationError::NotFound.new(
+                  "Attempted to insert relative to reference that does not exist #{ref}",
+                  [ref])
+              end
+
+              index += 1 if fupdate.after
+              target_indirect_viewmodels.insert(index, *appended_targets)
+            else
+              target_indirect_viewmodels.concat(appended_targets)
+            end
+
+          when FunctionalUpdate::Remove
+            removed_refs = fupdate.values.to_set
+            target_indirect_viewmodels.reject! { |target| removed_refs.include? target[:viewmodel_ref] }
+
+          when FunctionalUpdate::Update
+            # Just need to connect these through, so the ref is anchored somewhere in the update tree.
+            # Uniquess of vm_refs has already been checked.=
+            ref_by_indirect_vm_ref =
+              fupdate.values.index_by { |ref| update_context.resolve_reference(ref).viewmodel_reference }
+
+            target_indirect_viewmodels.each do |target|
+              associated_ref = ref_by_indirect_vm_ref.delete(target[:viewmodel_ref])
+              target[:ref]   = associated_ref if associated_ref
+            end
+
+            if ref_by_indirect_vm_ref.present?
+              raise ViewModel::DeserializationError::NotFound.new(
+                "Stale functional update for association '#{association_data.direct_reflection.name}' - "\
+                "could not match referenced viewmodels: [#{ref_by_indirect_vm_ref.keys.map(&:to_s).join(', ')}]")
+            end
+
+          else
+            raise ArgumentError.new("Unknown functional update: '#{fupdate.class}'")
+          end
+        end
+
+      else
+        raise_deserialization_error("Unknown association_update type '#{association_update.class.name}'")
+      end
+
+      new_through_viewmodels = target_indirect_viewmodels.map do |target|
+        target_viewmodel_ref, ref  = target.values_at(:viewmodel_ref, :ref)
 
         existing_through_viewmodel =
-          previous_through_viewmodels_by_indirect_ref[target_reference].try(:shift) if target_reference
+          (previous_through_viewmodels_by_indirect_ref[target_viewmodel_ref]&.shift if target_viewmodel_ref)
+
+        unless ref || existing_through_viewmodel
```

Comment:
so are we carrying around these pairs so that we can either construct or match existing? Would it be possible to instead match greedily rather than construct these intermediates?